src/raftstore/store/worker/mvcc_gc.rs

```diff
@@ -0,0 +1,219 @@
+use std::cmp::{max, min};
+use std::collections::{BTreeMap, HashMap};
+use std::fmt::{self, Display};
+use std::ops::Bound::Included;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::mpsc::{self, Receiver, SyncSender};
+use std::sync::Arc;
+use std::thread;
+use std::time::Duration;
+
+use crate::raftstore::store::keys::{data_end_key, data_key};
+use crate::storage::mvcc::{
+    extract_mvcc_props, get_gc_ratio, init_mvcc_gc_db, init_safe_point, need_gc,
+};
+use engine::rocks::util::get_cf_handle;
+use engine::rocks::{CompactOptions, DBBottommostLevelCompaction, DB};
+use engine::util::get_range_properties_cf;
+use engine::CF_WRITE;
+use tikv_util::time::{duration_to_ms, Instant};
+use tikv_util::timer::Timer;
+use tikv_util::worker::{Runnable, RunnableWithTimer};
+
+#[derive(Clone, Debug)]
+pub struct MvccGcTask {
+    pub safe_point: u64,
+    pub start_key: Vec<u8>,
+    pub end_key: Vec<u8>,
+}
+
+impl Display for MvccGcTask {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("MvccGc")
+            .field("safe_point", &self.safe_point)
+            .finish()
+    }
+}
+
+pub struct MvccGcRunner {
+    db: Arc<DB>,
+    safe_point: Arc<AtomicU64>,
+    tasks: BTreeMap<Vec<u8>, (Vec<u8>, usize)>,
+    running_tasks: HashMap<Vec<u8>, Vec<u8>>,
+    sender: SyncSender<(Vec<u8>, Vec<u8>)>,
+    receiver: Receiver<(Vec<u8>, Vec<u8>)>,
+}
+
+impl MvccGcRunner {
+    pub fn new(db: Arc<DB>) -> MvccGcRunner {
+        init_mvcc_gc_db(Arc::clone(&db));
+        let safe_point = Arc::new(AtomicU64::new(0));
+        init_safe_point(Arc::clone(&safe_point));
+
+        let (tx, rx) = mpsc::sync_channel(CONCURRENCY);
+        MvccGcRunner {
+            db,
+            safe_point,
+            tasks: BTreeMap::new(),
+            running_tasks: HashMap::new(),
+            sender: tx,
+            receiver: rx,
+        }
+    }
+
+    pub fn new_timer() -> Timer<()> {
+        let mut timer = Timer::new(2);
+        timer.add_task(COLLECT_TASK_INTERVAL, ());
+        timer
+    }
+}
+
+impl Runnable<MvccGcTask> for MvccGcRunner {
+    fn run(&mut self, task: MvccGcTask) {
+        self.safe_point.store(task.safe_point, Ordering::Release);
+        for (sk, ek) in &self.running_tasks {
+            let dsk = data_key(&task.start_key);
+            let dek = data_end_key(&task.end_key);
+            if *sk < dek && dsk < *ek {
+                // overlap with a range in compaction.
+                info!("gc task overlap, skip");
```

Comment:
Why not use a BTreeMap instead of a HashMap, so you don't need to iterate all ranges and check all of them?